src/AbstractUri.php

```diff
@@ -374,21 +374,52 @@ protected function formatHost($host)
             return strtolower($host);
         }
 
-        $component = '';
-        $valid_ascii_label_characters = '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ-';
-        foreach (explode('.', mb_strtolower($host, 'UTF-8')) as $label) {
-            if (false !== strpos($label, '%')) {
-                $label = rawurldecode($label);
-            }
+        $component = strtolower($host);
+        if (false !== strpos($component, '%')) {
+            $component = rawurldecode($component);
+        }
```

Comment:
I guess this could be moved above the `preg_match` check, but I'm not sure. Actually I wasn't aware host names could be encoded, where is this specified?

Reply:
it is written in [RFC3986](https://tools.ietf.org/html/rfc3986#section-3.5)

> reg-name    = *( unreserved / pct-encoded / sub-delims )
where pct-encoded is
>  pct-encoded = "%" HEXDIG HEXDIG

Reply:
Wrong section, but indeed, TIL. https://tools.ietf.org/html/rfc3986#section-3.2.2 is the right one. Most domains won't be encoded, especially if they're ASCII only, so let's keep the check below the regex. This is potentially slower if an ASCII only domain is encoded, but it makes the much more common case of no encoding and an ASCII only domain faster.

Reply:
This `if` should be moved below the `$reg_name` `preg_match`, because that will favor the happy path.